assertj-error-prone/src/main/java/com/palantir/assertj/errorprone/AssertjAssertThatThrownBy.java

```diff
@@ -0,0 +1,118 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.assertj.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.fixes.Fix;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.sun.source.tree.CatchTree;
+import com.sun.source.tree.ExpressionStatementTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.StatementTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.TryTree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.tree.JCTree;
+import java.util.List;
+import java.util.Optional;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "AssertjAssertThatThrownBy",
+        link = "https://github.com/palantir/assertj-automation",
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = BugPattern.SeverityLevel.SUGGESTION,
+        summary = "Prefer AssertJ assertThatThrownBy assertions over try-catch with fail statements")
+public final class AssertjAssertThatThrownBy extends BugChecker implements BugChecker.TryTreeMatcher {
+    private static final ImmutableList<String> IGNORED_FAIL_MESSAGES = ImmutableList.of("\"fail\"", "\"\"");
+
+    private static final Matcher<StatementTree> FAIL_METHOD = Matchers.anyOf(
+            Matchers.expressionStatement(
+                    MethodMatchers.staticMethod().onClass("org.junit.Assert").named("fail")),
+            Matchers.expressionStatement(MethodMatchers.staticMethod()
+                    .onClass("org.junit.jupiter.api.Assertions")
+                    .named("fail")),
+            Matchers.expressionStatement(MethodMatchers.staticMethod()
+                    .onClass("org.assertj.core.api.Assertions")
+                    .named("fail")));
+
+    @Override
+    public Description matchTry(TryTree tree, VisitorState state) {
+        List<? extends StatementTree> tryStatements = tree.getBlock().getStatements();
+        if (tryStatements.isEmpty() || tree.getCatches().size() != 1 || tree.getFinallyBlock() != null) {
+            return Description.NO_MATCH;
+        }
+        CatchTree catchTree = Iterables.getOnlyElement(tree.getCatches());
+        if (!catchTree.getBlock().getStatements().isEmpty()
```

Comment:
In a future improvement we can handle non-empty blocks using `isInstanceOfSatisfying(type, catch-block-as-lambda)`